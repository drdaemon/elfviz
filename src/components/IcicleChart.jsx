import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { hierarchy, partition } from 'd3-hierarchy';
import { scaleLinear, scaleOrdinal, schemeCategory20 } from 'd3-scale';

import ChartTooltip from './ChartTooltip';

class IcicleChart extends Component {

  constructor(props) {
    super(props);
    this.state = {
      tooltip: {
        display: false,
      },
    };

    this.color = scaleOrdinal(schemeCategory20);
    this.initializeScale(props);
  };

  componentWillReceiveProps(props) {
    this.initializeScale(props);
  }

  initializeScale(props) {
    const { width, height } = props;

    this.scaleX = scaleLinear().domain([0, width]).range([0, width]);
    this.scaleY = scaleLinear().domain([0, height]).range([0, height]);
  }

  showTooltip(tooltip = { display: false }, event) {
    this.setState({ tooltip });
  }

  onClickRect(d, event) {
    const { width, height } = this.props;

    this.scaleX.domain([d.x0 + 1, d.x1 - 1]);
    this.scaleY.domain([d.y0 - 1, height]).range([d.y0 > 1 ? 20 : 0, height]);

    this.forceUpdate(); // to trigger re-rendering, normally forceUpdate() should be avoided in the app
  }

  renderPartition() {
    const { data } = this.props;
    const rootData = {
      name: 'ELF',
      size: 0,
      children: data,
    };

    const root = hierarchy(rootData)
      .sum(d => d.size);

    const iciclePartition = partition()
      .size([this.props.width, this.props.height])
      .padding(1);

    const nodes = iciclePartition(root)
      .descendants();

    const rects = nodes.map((d, i) => {
      const rectStyle = {
        fill: this.color(i),
      };
      const x = this.scaleX(d.x0);
      const y = this.scaleY(d.y0);
      const width = this.scaleX(d.x1) - x;
      const height = this.scaleY(d.y1) - y;
      const textX = x + width / 2;
      const textY = y + height / 2;
      const transform = 'translate(' + textX + ',' + textY + ')rotate(90)';
      const tooltipObject = { // assigning object not very optimal, to be rewrited later
        data: {
          key: d.data.name,
          value: d.data.size,
        },
        display: true,
        pos: {
          x: textX,
          y: textY,
        },
      };

      return (
        <g key={'path' + i}>
          <rect
            className="node"
            x={x}
            y={y}
            width={width}
            height={height}
            style={rectStyle}
            onClick={this.onClickRect.bind(this, d)}
            onMouseOver={this.showTooltip.bind(this, tooltipObject)}
            onMouseOut={this.showTooltip.bind(this)}
          />
          {width > 6 ? <text className="label" dx="0.35em" transform={transform}>{d.data.name}</text> : null}
        </g>
      );
    });
    return rects;
  }
  
  render() {
    const { id, width, height } = this.props;
    const { tooltip } = this.state;
    const partition = this.renderPartition();
    return (
      <svg id={id} width={width} height={height}>
        {partition}
        {tooltip.display ? <ChartTooltip
          textStyle1="tooltip-text1"
          textStyle2="tooltip-text1"
          bgStyle="tooltip-bg"
          xValue="Section"
          yValue="Size"
          {...tooltip}
        /> : null}
      </svg>
    );
  }
}

IcicleChart.propTypes = {
  id: PropTypes.string.isRequired,
  data: PropTypes.array.isRequired,
  height: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
};

IcicleChart.defaultProps = {
  data: [],
  height: 0,
  width: 0,
};

export default IcicleChart;
