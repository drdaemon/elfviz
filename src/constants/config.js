// height of the visualization chart (svg)
export const height = 450;

export const qrcode = 'data:image/jpeg;base64,iVBORw0KGgoAAAANSUhEUgAAAXIAAAFyAQMAAADS6sNKAAAABlBMVEX///8AAABVwtN+AAABpElEQVR4nO2a0Y2EMAxELVEAJdE6JVEAUpbYniQb0N3fnQbNfHiBPPOD4zjOmkmSJEmS9D8qqTOubKvG1vMaWjG2i+fl4+YCrqvlQqvZqvvuTh0Rz8nXz+5j22HNLPEsAkX8K/gaCh4UMfNN/Kv4QMNEPhf/Bt5/2lg3v+Z/8Qx8KoEvgzHxvPzo2dZri8rsZ4ln4D0UosAutQpzs0fRXZDexfPyUVYf+TgNprqZeG4+J/jagmKIDHuox8Qz8rlBLjCez9cZFU/IGwowj4dsdSGfOzA5iafiazz0AEgnhEc0rcVT89mgRgAso8m3iaflR8BaKOR+2Z4knogfp3qrwnpT8xYP4rn4+YApluqWz2sOEE/M+94Jq3S2N813yeWxHhPPxXvFZYYCe9hUNX6SeCa+tz2m/VTqWIp4Yn7yjKDo7c17v0s8FY/vXs+G+8z3Kgz9LvHMfNysJyY9zhMzHm75QTwZj//zYKpjP1W+Zr74F/BYqvOMEau5+DfwSZXmfrRR8cS8/6CBObiP6V08LZ9qQEFk+Ivu/U/xVLwkSZIkSX+pDyPI6zwnAxz3AAAAAElFTkSuQmCC';
