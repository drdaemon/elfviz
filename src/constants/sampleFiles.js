import elfAC97KernelDriver from '../data/elfAC97KernelDriver.json';
import elfOpenSSLPHPModule from '../data/elfOpenSSLPHPModule.json';
import elfOpenSSLPHPModuleAfterStrip from '../data/elfOpenSSLPHPModule-after-strip.json';
import elfSimpleCProgram from '../data/elfSimpleCProgram.json';

export default {
  elfAC97KernelDriver,
  elfOpenSSLPHPModule,
  elfOpenSSLPHPModuleAfterStrip,
  elfSimpleCProgram,
};
