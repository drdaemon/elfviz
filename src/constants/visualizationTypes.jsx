import React from 'react';

export const visualizationTypes = [
  {
    name: 'Icicles',
    icon: <i className="fa fa-2x fa-bar-chart fa-rotate-180" aria-hidden="true"></i>,
  },
  {
    name: 'Sunburst',
    icon: <i className="fa fa-2x fa-pie-chart" aria-hidden="true"></i>,
  },
];
