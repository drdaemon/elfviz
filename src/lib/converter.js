export const convertReadElfToD3 = data => {
  return data.map(section => {
    const { name, size, symbols } = section;
    const item = { name, size };
    if (symbols) {
      item.children = symbols
        .filter(symbol => symbol.size > 0 && symbol.name !== '<none>')
        .map(symbol => {
          return {
            name: symbol.name,
            size: symbol.size,
          };
        });
    }
    // console.log('item', item);
    return item;
  });
};
