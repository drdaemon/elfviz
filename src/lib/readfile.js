export const readfile = fileToRead => {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = file => resolve(file.target.result);
    reader.onerror = err => reject(err);
    reader.readAsText(fileToRead, "UTF-8");
  });
};
