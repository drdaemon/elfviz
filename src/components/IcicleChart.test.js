import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import IcicleChart from './IcicleChart';

import * as config from '../constants/config';
import elfSimpleCProgram from '../data/elfSimpleCProgram';
import { convertReadElfToD3 } from '../lib/converter';

describe('IcicleChart', () => {
  it('renders without crashing', () => {
    shallow(<IcicleChart id="id" data={[]} height={0} width={0} />);
  });

  it('should match an empty snapshot', () => {
    const wrapper = shallow(<IcicleChart id="id" data={[]} height={0} width={0} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match a full data snapshot', () => {
    const data = convertReadElfToD3(elfSimpleCProgram);
    const wrapper = shallow(<IcicleChart id="id" data={data} height={config.height} width={500} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});
