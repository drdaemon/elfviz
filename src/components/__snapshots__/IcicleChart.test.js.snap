// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IcicleChart should match a full data snapshot 1`] = `
<svg
  height={450}
  id="id"
  width={500}
>
  <g>
    <rect
      className="node"
      height={148}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#1f77b4",
        }
      }
      width={498}
      x={1}
      y={1}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(250,75)rotate(90)"
    >
      ELF
    </text>
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#aec7e8",
        }
      }
      width={0}
      x={0.5}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#ff7f0e",
        }
      }
      width={1.7035603715170282}
      x={1}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#ffbb78",
        }
      }
      width={2.0897832817337463}
      x={3.703560371517028}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#2ca02c",
        }
      }
      width={2.4760061919504643}
      x={6.7933436532507745}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#98df8a",
        }
      }
      width={1.7035603715170282}
      x={10.269349845201239}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#d62728",
        }
      }
      width={17.538699690402474}
      x={12.972910216718267}
      y={150}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(21.742260061919502,224.5)rotate(90)"
    >
      .dynsym
    </text>
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#ff9896",
        }
      }
      width={9.331462848297214}
      x={31.511609907120743}
      y={150}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(36.177341331269346,224.5)rotate(90)"
    >
      .dynstr
    </text>
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#9467bd",
        }
      }
      width={0.5448916408668723}
      x={41.84307275541796}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#c5b0d5",
        }
      }
      width={3.6346749226006168}
      x={43.38796439628483}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#8c564b",
        }
      }
      width={1.317337461300312}
      x={48.022639318885446}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#c49c94",
        }
      }
      width={12.90402476780185}
      x={50.33997678018576}
      y={150}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(56.79198916408669,224.5)rotate(90)"
    >
      .rela.plt
    </text>
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#e377c2",
        }
      }
      width={1.510448916408663}
      x={64.24400154798761}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#f7b6d2",
        }
      }
      width={9.814241486068113}
      x={66.75445046439627}
      y={150}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(71.66157120743033,224.5)rotate(90)"
    >
      .plt
    </text>
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#7f7f7f",
        }
      }
      width={0}
      x={77.4549148606811}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#c7c7c7",
        }
      }
      width={48.629643962848306}
      x={78.34113777089782}
      y={150}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(102.65595975232196,224.5)rotate(90)"
    >
      .text
    </text>
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#bcbd22",
        }
      }
      width={0}
      x={127.90528250773991}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#dbdb8d",
        }
      }
      width={0.9311145510835956}
      x={128.83978328173373}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#17becf",
        }
      }
      width={4.020897832817326}
      x={130.77089783281733}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#9edae5",
        }
      }
      width={22.55959752321982}
      x={135.79179566563465}
      y={150}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(147.07159442724458,224.5)rotate(90)"
    >
      .eh_frame
    </text>
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#1f77b4",
        }
      }
      width={0}
      x={159.2376160990712}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#aec7e8",
        }
      }
      width={0}
      x={160.0100619195046}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#ff7f0e",
        }
      }
      width={0}
      x={160.78250773993807}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#ffbb78",
        }
      }
      width={43.80185758513932}
      x={161.66873065015477}
      y={150}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(183.56965944272443,224.5)rotate(90)"
    >
      .dynamic
    </text>
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#2ca02c",
        }
      }
      width={0}
      x={206.3568111455108}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#98df8a",
        }
      }
      width={5.952012383900922}
      x={207.24303405572752}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#d62728",
        }
      }
      width={0.5448916408668651}
      x={214.19504643962844}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#ff9896",
        }
      }
      width={0}
      x={215.62616099071204}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#9467bd",
        }
      }
      width={4.020897832817326}
      x={216.51238390092874}
      y={150}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#c5b0d5",
        }
      }
      width={24.87693498452012}
      x={221.53328173374607}
      y={150}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(233.97174922600612,224.5)rotate(90)"
    >
      .shstrtab
    </text>
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#8c564b",
        }
      }
      width={192.01489938080496}
      x={247.41021671826618}
      y={150}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(343.41766640866865,224.5)rotate(90)"
    >
      .symtab
    </text>
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#c49c94",
        }
      }
      width={58.574883900928796}
      x={440.42511609907115}
      y={150}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(469.7125580495356,224.5)rotate(90)"
    >
      .strtab
    </text>
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#e377c2",
        }
      }
      width={0}
      x={246.95849458204327}
      y={300}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#f7b6d2",
        }
      }
      width={0}
      x={247.10332817337454}
      y={300}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#7f7f7f",
        }
      }
      width={0}
      x={247.3929953560371}
      y={300}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#c7c7c7",
        }
      }
      width={8.752128482972125}
      x={248.08610681114544}
      y={300}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(252.4621710526315,374.5)rotate(90)"
    >
      __libc_csu_init
    </text>
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#bcbd22",
        }
      }
      width={3.0553405572755423}
      x={257.83823529411757}
      y={300}
    />
  </g>
  <g>
    <rect
      className="node"
      height={149}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#dbdb8d",
        }
      }
      width={13.000580495356019}
      x={261.8935758513931}
      y={300}
    />
    <text
      className="label"
      dx="0.35em"
      transform="translate(268.39386609907115,374.5)rotate(90)"
    >
      main
    </text>
  </g>
</svg>
`;

exports[`IcicleChart should match an empty snapshot 1`] = `
<svg
  height={0}
  id="id"
  width={0}
>
  <g>
    <rect
      className="node"
      height={0}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "fill": "#1f77b4",
        }
      }
      width={0}
      x={0}
      y={0}
    />
  </g>
</svg>
`;
